### ISSUE TITLE
RuntimeError: Pulse time after readout window end (Trac #1948)

### ISSUE THREAD

**Issue by icecube-trac** [icecube-trac/tix2#1940]
Below you can see one example the complete error message:

```text

FATAL (millipede): Pulse time (47860.000000+16.000000 ns) after readout window end (26400.000000 ns) on OM(3,2) (MillipedeDOMCacheMap.cxx:417 in void MillipedeDOMCacheMap::UpdateData(const I3TimeWindow&, const I3RecoPulseSeriesMap&, const I3TimeWindowSeriesMap&, double, double, double, bool))
ERROR (I3Module): Monopod16: Exception thrown (I3Module.cxx:116 in void I3Module::Do(void (I3Module::*)()))
Traceback (most recent call last):
  File "/home/lunger/data/meta-projects/processing/release/V05-00-00/trunk/src/WIMPprod/python/drivers/LevelX_complete.py", line 134, in <module>
    tray.Execute()
  File "/home/lunger/data/meta-projects/processing/release/V05-00-00/trunk/build/lib/I3Tray.py", line 234, in Execute
    super(I3Tray, self).Execute()
RuntimeError: Pulse time (47860.000000+16.000000 ns) after readout window end (26400.000000 ns) on OM(3,2) (in void MillipedeDOMCacheMap::UpdateData(const I3TimeWindow&, const I3RecoPulseSeriesMap&, const I3TimeWindowSeriesMap&, double, double, double, bool))

 *** Break *** bus error
```

If you would like to see the rest you can take a look at e.g.: 

```text
/data/ana/BSM/HT_Cascade/processed/2013/L3_wL4var/corsika/12161/logs/L3_wL4var-corsika.012161.007415.err
```

I am using I3WaveformTimeRangeCalculator in the tray, and I tried the function from jvs: 

```text
def fix_bad_timeshift(frame,Pulses):

        trigs = frame['I3TriggerHierarchy']

        if trigs[0].time != 0:

            twindow_name = Pulses+'TimeRange'

            twindow = frame[twindow_name]

            del frame[twindow_name]

            frame[twindow_name] = dataclasses.I3TimeWindow(twindow.start + trigs[0].time, twindow.stop + trigs[0].time)
```

The error still remains. You can find the tray script here:

```text
/home/lunger/meta-projects/processing/release/V05-00-00/trunk/src/WIMPprod/python/trays/L3_wL4var.py
```

Migrated from https://code.icecube.wisc.edu/ticket/1948
```json
{
    "status": "closed", 
    "changetime": "2019-02-13T14:13:32", 
    "description": "Below you can see one example the complete error message:\n\n{{{\n\nFATAL (millipede): Pulse time (47860.000000+16.000000 ns) after readout window end (26400.000000 ns) on OM(3,2) (MillipedeDOMCacheMap.cxx:417 in void MillipedeDOMCacheMap::UpdateData(const I3TimeWindow&, const I3RecoPulseSeriesMap&, const I3TimeWindowSeriesMap&, double, double, double, bool))\nERROR (I3Module): Monopod16: Exception thrown (I3Module.cxx:116 in void I3Module::Do(void (I3Module::*)()))\nTraceback (most recent call last):\n  File \"/home/lunger/data/meta-projects/processing/release/V05-00-00/trunk/src/WIMPprod/python/drivers/LevelX_complete.py\", line 134, in <module>\n    tray.Execute()\n  File \"/home/lunger/data/meta-projects/processing/release/V05-00-00/trunk/build/lib/I3Tray.py\", line 234, in Execute\n    super(I3Tray, self).Execute()\nRuntimeError: Pulse time (47860.000000+16.000000 ns) after readout window end (26400.000000 ns) on OM(3,2) (in void MillipedeDOMCacheMap::UpdateData(const I3TimeWindow&, const I3RecoPulseSeriesMap&, const I3TimeWindowSeriesMap&, double, double, double, bool))\n\n *** Break *** bus error\n}}}\n\n\n\nIf you would like to see the rest you can take a look at e.g.: \n\n\n{{{\n/data/ana/BSM/HT_Cascade/processed/2013/L3_wL4var/corsika/12161/logs/L3_wL4var-corsika.012161.007415.err\n}}}\n\nI am using I3WaveformTimeRangeCalculator in the tray, and I tried the function from jvs: \n\n{{{\ndef fix_bad_timeshift(frame,Pulses):\n\n        trigs = frame['I3TriggerHierarchy']\n\n        if trigs[0].time != 0:\n\n            twindow_name = Pulses+'TimeRange'\n\n            twindow = frame[twindow_name]\n\n            del frame[twindow_name]\n\n            frame[twindow_name] = dataclasses.I3TimeWindow(twindow.start + trigs[0].time, twindow.stop + trigs[0].time)\n}}}\n\nThe error still remains. You can find the tray script here:\n\n{{{\n/home/lunger/meta-projects/processing/release/V05-00-00/trunk/src/WIMPprod/python/trays/L3_wL4var.py\n}}}\n", 
    "reporter": "lisa", 
    "cc": "claudio.kopper, kjmeagher, olivas, juancarlos", 
    "resolution": "fixed", 
    "_ts": "1550067212810271", 
    "component": "combo simulation", 
    "summary": "RuntimeError: Pulse time after readout window end", 
    "priority": "blocker", 
    "keywords": "", 
    "time": "2017-02-10T00:22:04", 
    "milestone": "", 
    "owner": "jbraun", 
    "type": "defect"
}
```

**Comment by icecube-trac** [comment]
Trac update at 20170211T00:52:10: 
* olivas changed milestone from "" to "February 2017"
* olivas changed cc from "michael.gustafsson@theorie.physik.uni-goettingen.de, celaisat@gmail.com" to "claudio.kopper, kjmeagher, jbraun, juancarlos"
* olivas changed status from "new" to "accepted"
* olivas changed owner from "" to "olivas"

**Comment by icecube-trac** [comment]
Trac update at 20170228T16:50:01: 
* olivas changed milestone from "February 2017" to "March 2017"
* olivas changed cc from "claudio.kopper, kjmeagher, jbraun, juancarlos" to "claudio.kopper, kjmeagher, olivas, juancarlos"
* olivas changed owner from "olivas" to "jbraun"
* olivas changed status from "accepted" to "assigned"

**Comment by icecube-trac** [comment]
Trac update at 20170313T22:25:24: 
* olivas commented:

> Upping this to 'blocker' before we get the next release out.  We can always downgrade, but we shouldn't release until this is investigated further.


* olivas changed priority from "major" to "blocker"

**Comment by icecube-trac** [comment]
Trac update at 20170316T19:37:22: 
* david.schultz changed resolution from "" to "worksforme"
* david.schultz changed status from "assigned" to "closed"
* david.schultz commented:

> This is actually a bug in your own script.  When calling millipede, make sure you give it the right ReadoutWindow for the Pulses you are running on.
> 
> This crashes with the error mentioned in the ticket:
> ``` 
> Pulses = uncleaned_recopulse,
> ReadoutWindow = "CalibratedWaveformRange"
> ```
> 
> This works just fine:
> ``` 
> Pulses = uncleaned_recopulse,
> ReadoutWindow = uncleaned_recopulse+'TimeRange'
> ```

**Comment by icecube-trac** [comment]
Trac update at 20170317T18:35:46: 
* jbraun commented:

> Nancy W. has identified events from icesim5 with pulses in OfflinePulses outside of OfflinePulsesTimeRange.  This is a bigger issue, so I'm reopening. 


* jbraun changed resolution from "worksforme" to ""
* jbraun changed status from "closed" to "reopened"

**Comment by icecube-trac** [comment]
Trac update at 20170320T17:00:56: jbraun commented:

> The time range in the I3EventHeader is used to determine CalibratedWaveformsRange (and therefore OfflinePulsesTimeRange), so we assume that the event start/stop time window contains the launch times.
> 
> In icesim5, I3GlobalTriggerSim.cxx, the I3EventHeader time window is not set correctly.  The end_time calculation does not subtract the earliest hit/trigger time, resulting in a window that is too small/large if the earliest time is negative/positive, respectively.  This affects all of icesim5 simulation.

**Comment by icecube-trac** [comment]
Trac update at 20170320T17:59:56: 
* david.schultz changed status from "reopened" to "closed"
* david.schultz commented:

> In changeset 154010:
> 
> ```text
> #!CommitTicketReference repository="IceCube" revision="154010"
> fix I3EventHeader end times. verify with a unit test. fixes #1940
> ```


* david.schultz changed resolution from "" to "fixed"

**Comment by icecube-trac** [comment]
Trac update at 20190213T14:13:32: 
* olivas commented:

> Milestone March 2017 deleted


* olivas changed milestone from "March 2017" to ""
